deeppavlov/pipeline_manager/pipeline_manager.py

```diff
@@ -0,0 +1,519 @@
+        self.sample_num = self.exp_config['pipeline_search'].get('sample_num', 10)
+        self.target_metric = self.exp_config['pipeline_search'].get('target_metric')
+        self.multiprocessing = self.exp_config['pipeline_search'].get('multiprocessing', True)
+        self.max_num_workers = self.exp_config['pipeline_search'].get('max_num_workers')
+        cpu_num = cpu_count()
+        if self.max_num_workers:
+            if self.max_num_workers > cpu_num:
+                raise ConfigError("Parameter 'max_num_workers'={0}, "
+                                  "but amounts of cpu is {1}.".format(self.max_num_workers, cpu_num))
+            elif self.max_num_workers < 1:
+                raise ConfigError("The number of workers must be at least equal to one. "
+                                  "Please check 'max_num_workers' parameter in config.")
+
+        self.use_gpu = self.exp_config['pipeline_search'].get('use_all_gpus', False)
+        self.memory_fraction = self.exp_config['pipeline_search'].get('gpu_memory_fraction', 1.0)
+        self.max_num_workers = None
+        self.available_gpu = None
+
+        # create the observer
+        self.save_path = self.root / self.date / self.exp_name / 'checkpoints'
+        self.observer = ExperimentObserver(self.exp_name, self.root, self.info, self.date, self.plot)
+        # create the pipeline generator
+        self.pipeline_generator = PipeGen(self.exp_config, self.save_path, self.search_type, self.sample_num, False)
+        self.gen_len = self.pipeline_generator.length
+        # write train data in observer
+        self.observer.exp_info['number_of_pipes'] = copy(self.gen_len)
+        self.observer.exp_info['dataset_name'] = copy(self.exp_config['dataset_reader']['data_path'].split("/")[-1])
+
+        self.observer.exp_info['metrics'] = []
+        for met in copy(self.exp_config['train']['metrics']):
+            if isinstance(met, dict):
+                self.observer.exp_info['metrics'].append(met['name'])
+            else:
+                self.observer.exp_info['metrics'].append(met)
+
+        if self.target_metric:
+            self.observer.exp_info['target_metric'] = self.target_metric
+        else:
+            self.observer.exp_info['target_metric'] = self.observer.exp_info['metrics'][0]
+
+        self.prepare_multiprocess()
+
+        # write time of experiment start
+        self.start_exp = time.time()
+        # start test
+        if self.do_test:
+            self.test()
+
+    def prepare_multiprocess(self) -> None:
+        """
+        Calculates the number of workers and the set of available video cards, if gpu is used, based on init attributes.
+        """
+        try:
+            visible_gpu = [int(q) for q in os.environ['CUDA_VISIBLE_DEVICES'].split(',')]
+            os.environ['CUDA_VISIBLE_DEVICES'] = ""
+        except KeyError:
+            visible_gpu = None
+
+        if self.use_gpu:
+            if isinstance(self.use_gpu, (List[int], int)):
```

Comment:
`TypeError: Parameterized generics cannot be used with class or instance checks`